pkg/vmprovider/providers/vsphere2/session/session_vm_update.go

```diff
@@ -777,32 +782,206 @@ func (s *Session) attachClusterModule(
 	return s.Client.ClusterModuleClient().AddMoRefToModule(vmCtx, moduleUUID, resVM.MoRef())
 }
 
+func (s *Session) poweredOffVirtualMachine(
+	vmCtx context.VirtualMachineContextA2,
+	vcVM *object.VirtualMachine,
+	resVM *res.VirtualMachine,
+	moVM *mo.VirtualMachine,
+	existingPowerState vmopv1.VirtualMachinePowerState) (refetchProps bool, err error) {
+
+	var powerOff bool
+	if existingPowerState == vmopv1.VirtualMachinePowerStateOn {
+		powerOff = true
+	} else if existingPowerState == vmopv1.VirtualMachinePowerStateSuspended {
+		if vmCtx.VM.Spec.PowerOffMode == vmopv1.VirtualMachinePowerOpModeHard {
+			powerOff = true
+		}
+	}
+	if powerOff {
+		err = resVM.SetPowerState(
+			logr.NewContext(vmCtx, vmCtx.Logger),
+			existingPowerState,
+			vmCtx.VM.Spec.PowerState,
+			vmCtx.VM.Spec.PowerOffMode)
+		if err != nil {
+			return
+		}
+
+		refetchProps = true
+	}
+
+	// A VM's hardware can only be upgraded if the VM is powered off.
+	// var opResult vmutil.ReconcileMinHardwareVersionResult
+	opResult, err := vmutil.ReconcileMinHardwareVersion(
+		vmCtx,
+		vcVM.Client(),
+		*moVM,
+		false,
+		vmCtx.VM.Spec.MinHardwareVersion)
+	if err != nil {
+		return
+	}
+	if opResult == vmutil.ReconcileMinHardwareVersionResultUpgraded {
+		refetchProps = true
+	}
+
+	return refetchProps, err
+}
+
+func (s *Session) suspendedVirtualMachine(
```

Comment:
This is a very confusing function name. Can you please rename it to be something more descriptive? Or add a comment?